src/App.tsx

```diff
@@ -51,15 +53,18 @@ const UnauthenticatedRoute = ({
 );
 
 const App = (): JSX.Element => {
-  const [sessionUser, setSessionUser] = useState<User | null>(null);
-  const [hasSessionLoaded, setSessionLoaded] = useState(false);
+  const hasSessionLoaded = useSelector(
+    (state: GlobalState) => state.hasSessionLoaded
+  );
+  const sessionUser = useSelector((state: GlobalState) => state.sessionUser);
+  const dispatch = useDispatch();
 
   useEffect(() => {
     fetch("/session")
       .then(response => response.json())
       .then((data: { user: User | null }) => {
-        setSessionUser(data.user);
-        setSessionLoaded(true);
+        dispatch(setSessionLoaded(true));
```

Comment:
One note about action design - usually, if I see that we're dispatching multiple times in a row, it feels like an antipattern to me, because that causes multiple renders. It can be helpful to think of actions as _descriptions of what happened_ with relevant information from the thing that happened, but agnostic to what should change as a result, and let the store decide if/how to update from that. Here, we're being more prescriptive, telling the store how it should respond to the changes. Instead, we could do `dispatch(fetchedSession(data.user))`, and let the reducer take care of knowing that it should set the `sessionLoaded` flag to true as a result.

It ends up with less reuse of the actions, but makes for cleaner rerenders based on server interactions. It also centers some logic that we might want to be true for a bunch of cases, like loading state - if we wanted to show a spinner and then hide it when the loading was done, but keep the `sessionLoaded = false` if the request failed, now we can handle that in the reducer, instead of in every component that needs to do data fetching.